pkg/controllers/expiration/controller.go

```diff
@@ -0,0 +1,123 @@
+	kubeClient client.Client
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		kubeClient: kubeClient,
+	}
+}
+
+// Reconcile executes an expiration control loop for a node
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	// 1. Get node
+	node := &v1.Node{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// 2. Ignore if node is already deleting
+	if !node.DeletionTimestamp.IsZero() {
+		return reconcile.Result{}, nil
+	}
+	// 3. Ignore if provisioner doesn't exist
+	name, ok := node.Labels[v1alpha2.ProvisionerNameLabelKey]
+	if !ok {
+		return reconcile.Result{}, nil
+	}
+	namespace, ok := node.Labels[v1alpha2.ProvisionerNamespaceLabelKey]
+	if !ok {
+		return reconcile.Result{}, nil
+	}
+	provisioner := &v1alpha2.Provisioner{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, provisioner); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// 4. Ignore if TTLSecondsUntilExpired isn't defined
+	if provisioner.Spec.TTLSecondsUntilExpired == nil {
+		return reconcile.Result{}, nil
+	}
+	// 5. Trigger termination workflow if expired
+	expirationTTL := time.Duration(ptr.Int64Value(provisioner.Spec.TTLSecondsUntilExpired)) * time.Second
+	expirationTime := node.CreationTimestamp.Add(expirationTTL)
+	if time.Now().After(expirationTime) {
+		zap.S().Infof("Triggering termination after %s (+%s) for expired node %s", expirationTTL, time.Since(expirationTime), node.Name)
+		if err := c.kubeClient.Delete(ctx, node); err != nil {
+			return reconcile.Result{}, fmt.Errorf("expiring node %s, %w", node.Name, err)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// 6. Backoff until expired
+	return reconcile.Result{RequeueAfter: time.Until(expirationTime)}, nil
+}
+
+func (c *Controller) provisionerToNodes(o client.Object) (requests []reconcile.Request) {
+	nodes := &v1.NodeList{}
+	if err := c.kubeClient.List(context.Background(), nodes, client.MatchingLabels(map[string]string{
+		v1alpha2.ProvisionerNameLabelKey:      o.GetName(),
+		v1alpha2.ProvisionerNamespaceLabelKey: o.GetNamespace(),
+	})); err != nil {
+		zap.S().Errorf("Failed to list nodes when mapping expiration watch events, %s", err.Error())
+	}
+	for _, node := range nodes.Items {
+		requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
+	}
+	return requests
+}
+
+func (c *Controller) Register(m manager.Manager) error {
+	return controllerruntime.
+		NewControllerManagedBy(m).
+		Named("Expiration").
+		For(&v1.Node{}).
+		Watches(
+			// Reconcile all nodes related to a provisioner when it changes.
+			&source.Kind{Type: &v1alpha2.Provisioner{}},
```

Comment:
Just double-checking ... does this work (well) if there are multiple `Provisioner`s in the same cluster? (Say there are 3 Provisioners, I assume this means each one will do the fan-out any time one of them changes?) (Not necessarily a big deal, just making sure I understand).

Reply:
It won't fully fan out, since we're constraining nodes that are "owned" by a provisioner on line 101/102. We're effectively mapping provisioners to nodes and will trigger a reconcile loop for all nodes whenever a provisioner changes. 

e.g. Provisioner 1 updates it's TTL, all nodes for that provisioner will reconcile to check that TTL.

Reply:
But won't all 3 Provisioners reconcile (because `For(&v1.Node{})` on line 116?) and do the fan-out?

Reply:
For and Watches are two interfaces (at different levels of granularity) that establish watches. There is no coupling between the calls.

1. The For(Nodes) statement says that any changes to nodes will trigger an expiration reconcile loop for that node. This is *slightly* overkill. We could probably limit this to just update, and limit it to trigger an event if it has a provisioner w/ ttluntilexpired, but instead of putting those checks in the event handler, we just put them into the reconcile loop. The theory is that with ctrl batching, the performance is likely slightly better this way (less logic on the event batching side). 
2. The Watches(Provisioners, MapFunc(Provisioner -> Nodes)) says that any changes to provisioners will trigger an expiration reconcile loop for a set of nodes. This is again probably overkill -- could be applied on just create/update (not delete). The same logic applies for putting this logic on the event handling side vs the reconcile side. 

Probably easier to discuss in person if I'm still missing something.

Reply:
I don't think it's a blocking issue (probably just my misunderstanding), let's discuss offline.